File: app-kotlin/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/activities/investigation/evidence/EvidenceFragment.java
package com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.FrameLayout;
import android.widget.PopupWindow;
import android.widget.RelativeLayout;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.tritiumstudios.phasmophobiaevidencetool.R;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.views.SanityMeterView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.views.evidence.EvidenceList;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.views.ghost.GhostList;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.views.investigation.InvestigationSection;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.InvestigationFragment;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.DifficultyCarouselView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.MapCarouselView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.PhaseTimerControlView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.PhaseTimerView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.SanitySeekBarView;
import com.tritiumstudios.phasmophobiaevidencetool.activities.investigation.evidence.children.solo.views.WarnTextView;
import com.tritiumstudios.phasmophobiaevidencetool.listeners.CompositeListener;

/**
 * EvidenceFragment class
 *
 * @author TritiumGamingStudios
 */
public class EvidenceFragment extends InvestigationFragment {

    protected InvestigationSection ghostSection, evidenceSection;

    protected GhostList ghostList;
    protected EvidenceList evidenceList;

    protected ConstraintLayout sanityTrackingConstraintLayout;

    protected AppCompatImageView toggleSanityButton;
    protected AppCompatTextView phaseTimerTextView;
    protected AppCompatTextView sanityPercentTextView;

    protected CompositeListener compositeListenerPrev, compositeListenerNext;

    protected DifficultyCarouselView difficultyCarouselView;
    protected PhaseTimerView phaseTimerCountdownView;
    protected PhaseTimerControlView playPauseButton;
    protected MapCarouselView mapTrackControl;
    protected SanitySeekBarView sanitySeekBarView;
    protected SanityMeterView sanityMeterView;
    protected WarnTextView sanityWarningTextView;


    /**
     * EvidenceFragment constructor
     * <p>
     * TODO
     *
     * @param layout -
     */
    public EvidenceFragment(int layout) {
        super(layout);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_evidence_solo, container, false);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        super.onViewCreated(view, savedInstanceState);

        if(evidenceViewModel.getSanityData() != null) {
            evidenceViewModel.getSanityData()
                    .setFlashTimeoutMax(globalPreferencesViewModel.getHuntWarningFlashTimeout());
        }

        // GHOST / EVIDENCE CONTAINERS
        FrameLayout
                column_left = view.findViewById(R.id.column_left),
                column_right = view.findViewById(R.id.column_right);
        column_right.findViewById(R.id.scrollview)
                .setVerticalScrollbarPosition(View.SCROLLBAR_POSITION_RIGHT);

        if(!globalPreferencesViewModel.getIsLeftHandSupportEnabled()) {
            ghostSection = (InvestigationSection) column_left.getChildAt(0);
            evidenceSection = (InvestigationSection) column_right.getChildAt(0);
        } else {
            evidenceSection = (InvestigationSection) column_left.getChildAt(0);
            ghostSection = (InvestigationSection) column_right.getChildAt(0);
        }

        ghostSection.setLabel(getString(R.string.investigation_section_title_ghosts));
        evidenceSection.setLabel(getString(R.string.investigation_section_title_evidence));

        ScrollView ghost_scrollview = ghostSection.findViewById(R.id.scrollview);
        ScrollView evidence_scrollview = evidenceSection.findViewById(R.id.scrollview);

        ghostList = new GhostList(getContext());
        evidenceList = new EvidenceList(getContext());

        ghostList.init(evidenceViewModel, popupWindow,
                ghostSection.findViewById(R.id.progressbar),
                adRequest);
        evidenceList.init(evidenceViewModel, popupWindow,
                evidenceSection.findViewById(R.id.progressbar),
                adRequest, ghostList);

        ViewStub list_ghosts = ghostSection.findViewById(R.id.list);
        ViewStub list_evidence = evidenceSection.findViewById(R.id.list);

        ghost_scrollview.removeView(list_ghosts);
        ghost_scrollview.addView(ghostList);

        evidence_scrollview.removeView(list_evidence);
        evidence_scrollview.addView(evidenceList);

        //ghostContainer.requestDisallowInterceptTouchEvent(true);

        toggleSanityButton = view.findViewById(R.id.button_toggleSanity);

        // TIMER VIEW
        phaseTimerTextView = view.findViewById(R.id.evidence_timer_text);

        // SANITY METER VIEWS
        sanityPercentTextView = view.findViewById(R.id.evidence_sanitymeter_percentage);
        sanityMeterView = view.findViewById(R.id.evidence_sanitymeter_progressbar);
        sanitySeekBarView = view.findViewById(R.id.evidence_sanitymeter_seekbar);
        sanityWarningTextView = view.findViewById(R.id.evidence_sanitymeter_huntwarning);

        // SANITY COLLAPSIBLE
        sanityTrackingConstraintLayout = view.findViewById(R.id.constraintLayout_sanityTracking);

        if(toggleSanityButton != null) {
            toggleSanityButton.setOnClickListener(v -> {
                if(evidenceViewModel.isCollapsed()) {
                    sanityTrackingConstraintLayout.animate()
                            .setListener(new AnimatorListenerAdapter() {
                                @Override
                                public void onAnimationStart(Animator animation) {
                                    super.onAnimationStart(animation);
                                    sanityTrackingConstraintLayout.setVisibility(View.GONE);
                                }

                                @Override
                                public void onAnimationEnd(Animator animation) {
                                    super.onAnimationEnd(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
                                    toggleSanityButton.setImageLevel(2);
                                    evidenceViewModel.setCollapsed(false);
                                }
                            })
                            .start();
                } else {
                    sanityTrackingConstraintLayout.animate()
                            .setListener(new AnimatorListenerAdapter() {
                                @Override
                                public void onAnimationStart(Animator animation) {
                                    super.onAnimationStart(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
                                }
                                @Override
                                public void onAnimationEnd(Animator animation) {
                                    super.onAnimationStart(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.GONE);
                                    toggleSanityButton.setImageLevel(1);
                                    evidenceViewModel.setCollapsed(true);
                                }
                            })
                            .start();
                }
            });

            initCollapsible();
        }

        // SANITY
        if (sanitySeekBarView != null) {
            sanitySeekBarView.init(
                    evidenceViewModel.getSanityData(),
                    sanityPercentTextView);
            sanitySeekBarView.resetProgress();
        }

        sanityMeterView.init(evidenceViewModel.getSanityData());

        popupWindow = new PopupWindow(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT);

        new Thread(() -> ghostList.createGhostViews(popupWindow)).start();
        new Thread(() -> evidenceList.createEvidenceViews(popupWindow)).start();

    }

    private void initCollapsible() {
        if(evidenceViewModel.isCollapsed()) {
            sanityTrackingConstraintLayout.setVisibility(View.GONE);
            toggleSanityButton.setImageLevel(1);
        } else {
            sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
            toggleSanityButton.setImageLevel(2);
        }
    }

    @Override
    public void softReset() {
        if (evidenceViewModel != null) {
            evidenceViewModel.reset();
        }

        if (sanitySeekBarView != null) {
            sanitySeekBarView.updateProgress();
        }

        if(phaseTimerCountdownView != null) {
            phaseTimerCountdownView.destroyTimer();
        }
    }


    public void requestInvalidateComponents() {

        if(evidenceViewModel != null) {
            evidenceViewModel.getGhostOrderData().updateOrder();
        }

        if(evidenceList != null) {
            evidenceList.forceResetEvidenceContainer();
        }

        if(ghostList != null) {
            ghostList.forceResetGhostContainer();
        }

        // SANITY
        if (sanitySeekBarView != null) {
            sanitySeekBarView.resetProgress();
        }

        if(evidenceViewModel != null &&
                evidenceViewModel.getPhaseTimerData() != null) {
            playPauseButton.pause();
            phaseTimerCountdownView.reset();
        }

        if(sanityWarningTextView != null) {
            sanityWarningTextView.reset();
        }

    }

    /**
     * onPause
     */
    @Override
    public void onPause() {

        phaseTimerCountdownView.destroyTimer();

        super.onPause();
    }

    /**
     * onDestroyView
     */
    @Override
    public void onDestroyView() {

        if (sanityMeterView != null) {
            sanityMeterView.recycleBitmaps();
        }

        if(popupWindow != null) {
            popupWindow.dismiss();
            popupWindow = null;
        }

        super.onDestroyView();
    }

    /**
     * onResume
     */
    @Override
    public void onResume() {

        if (!sanityMeterView.hasBuiltImages()) {
            sanityMeterView.buildImages();
        }

        super.onResume();
    }

    @Override
    protected void saveStates() {

    }

}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/investigation/evidence/EvidenceFragment.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.PopupWindow
import android.widget.RelativeLayout
import android.widget.ScrollView
import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.ComposeView
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.InvestigationFragment
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence.views.investigation.InvestigationSectionWrapper
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence.views.investigation.sanity.SanityToolsLayout
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence.views.investigation.section.InvestigationSection
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence.views.investigation.section.lists.EvidenceListView
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.evidence.views.investigation.section.lists.GhostListView
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.compose.CollapseButton
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.compose.InvestigationToolbar
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.compose.ResetButton
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.compose.ToolBarItemPair
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

open class EvidenceFragment(layout: Int) : InvestigationFragment(layout) {
    private var investigationSectionWrapper: InvestigationSectionWrapper? = null

    private var ghostSection: InvestigationSection? = null
    private var evidenceSection: InvestigationSection? = null

    private var ghostList: GhostListView? = null
    private var evidenceList: EvidenceListView? = null

    protected var sanityToolsLayout: SanityToolsLayout? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_evidence, container, false)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        sanityToolsLayout = view.findViewById(R.id.layout_sanity_tool)
        investigationViewModel?.let { investigationViewModel ->
            sanityToolsLayout?.init(investigationViewModel)
        }

        // GHOST / EVIDENCE CONTAINERS
        investigationSectionWrapper = view.findViewById(R.id.layout_evidence_tool_wrapper)
        val columnLeft = investigationSectionWrapper?.findViewById<FrameLayout>(R.id.column_left)
        val columnRight = investigationSectionWrapper?.findViewById<FrameLayout>(R.id.column_right)
        columnRight?.findViewById<View>(R.id.scrollview)?.verticalScrollbarPosition =
            View.SCROLLBAR_POSITION_RIGHT

        globalPreferencesViewModel?.let { globalPreferencesViewModel ->
            if (!globalPreferencesViewModel.isLeftHandSupportEnabled) {
                ghostSection = columnLeft?.getChildAt(0) as InvestigationSection
                evidenceSection = columnRight?.getChildAt(0) as InvestigationSection
            } else {
                evidenceSection = columnLeft?.getChildAt(0) as InvestigationSection
                ghostSection = columnRight?.getChildAt(0) as InvestigationSection
            }
        }

        ghostSection?.setLabel(getString(com.tritiumgaming.core.resources.R.string.investigation_section_title_ghosts))
        evidenceSection?.setLabel(getString(com.tritiumgaming.core.resources.R.string.investigation_section_title_evidence))

        ghostList = GhostListView(requireContext())
        evidenceList = EvidenceListView(requireContext())

        val setupLists = CoroutineScope(Dispatchers.Main).launch {
            globalPreferencesViewModel?.let { globalPreferencesViewModel ->
                investigationViewModel?.let { investigationViewModel ->
                    evidenceList?.init(globalPreferencesViewModel, investigationViewModel,
                        popupWindow, evidenceSection?.findViewById(R.id.progressbar),
                        adRequest, ghostList)
                    ghostList?.init(globalPreferencesViewModel, investigationViewModel,
                        popupWindow, ghostSection?.findViewById(R.id.progressbar), adRequest)
                }
            }
        }
        setupLists.start()

        val ghostScrollview = ghostSection?.findViewById<ScrollView>(R.id.scrollview)
        val evidenceScrollview = evidenceSection?.findViewById<ScrollView>(R.id.scrollview)

        ghostScrollview?.addView(ghostList)
        evidenceScrollview?.addView(evidenceList)

        /*val buttonReset = view.findViewById<ComposeView?>(R.id.button_reset)
        initResetButton(buttonReset)*/

        val collapseButton: @Composable () -> Unit = {
            investigationViewModel?.investigationModel?.let { investigationModel ->

                CollapseButton(
                    isCollapsedState = investigationModel.isSanityDrawerCollapsed,
                    onClick = { investigationModel.toggleDrawerState() }
                ) { state ->
                    when (state) {
                        true ->
                            sanityToolsLayout?.animate()
                                ?.setListener(object : AnimatorListenerAdapter() {
                                    override fun onAnimationStart(animation: Animator) {
                                        super.onAnimationStart(animation)
                                        sanityToolsLayout?.visibility = View.GONE
                                    }
                                    override fun onAnimationEnd(animation: Animator) {
                                        super.onAnimationEnd(animation)
                                        sanityToolsLayout?.visibility = View.VISIBLE
                                    }
                                })?.start()

                        false -> {
                            sanityToolsLayout?.animate()
                                ?.setListener(object : AnimatorListenerAdapter() {
                                    override fun onAnimationStart(animation: Animator) {
                                        super.onAnimationStart(animation)
                                        sanityToolsLayout?.visibility = View.VISIBLE
                                    }

                                    override fun onAnimationEnd(animation: Animator) {
                                        super.onAnimationStart(animation)
                                        sanityToolsLayout?.visibility = View.GONE
                                    }
                                })?.start()
                        }
                    }
                }

            }
        }

        val resetButton: @Composable () -> Unit = {
            ResetButton(
                onClick = { reset() }
            )
        }

        val toolsList: Array<ToolBarItemPair> = arrayOf(
            ToolBarItemPair(collapseButton),
            ToolBarItemPair(resetButton),
            ToolBarItemPair(View(context))
        )

        val toolbarComposable: @Composable () -> Unit = {
            InvestigationToolbar(
            toolsList
            )
        }

        val sanityToolbarComposable: ComposeView? =
            view.findViewById(R.id.sanityToolbarComposable)
        sanityToolbarComposable?.setContent { toolbarComposable() }

        popupWindow = PopupWindow(
            RelativeLayout.LayoutParams.MATCH_PARENT,
            RelativeLayout.LayoutParams.MATCH_PARENT
        )

        ghostList?.createPopupWindow(popupWindow)
        evidenceList?.createPopupWindow(popupWindow)

        Thread { ghostList?.createViews() }.start()
        Thread { evidenceList?.createViews() }.start()

    }

    override fun reset() {
        investigationViewModel?.reset()
        objectivesViewModel?.reset()
        investigationViewModel?.investigationModel?.ghostOrderModel?.updateOrder()
        ghostList?.reset()
    }

    override fun onDestroyView() {
        popupWindow?.dismiss()
        popupWindow = null

        super.onDestroyView()
    }

    override fun saveStates() { }

}
